src/interface/swarm.hpp

```diff
@@ -0,0 +1,150 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_HPP_
+#define INTERFACE_SWARM_HPP_
+
+///
+/// A swarm contains all particles of a particular species
+/// Date: August 21, 2019
+
+#include "basic_types.hpp"
+#include "bvals/cc/bvals_cc.hpp"
+#include "metadata.hpp"
+#include "parthenon_arrays.hpp"
+#include "variable.hpp"
+#include <array>
+#include <cstdint>
+#include <list>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace parthenon {
+class MeshBlock;
+
+enum class PARTICLE_STATUS { UNALLOCATED, ALIVE, DEAD };
+
+class Swarm {
+ public:
+  MeshBlock *pmy_block = nullptr; // ptr to MeshBlock
+
+  Swarm(const std::string label, const Metadata &metadata, const int nmax_pool_in = 3);
+
+  // TODO BRR This should really be const... mask_ is managed internally
+  /// Get mask array for active particles
+  ParticleVariable<bool> &GetMask() { return mask_; }
+
+  /// Get the mask array of active particles that are marked for removal
+  ParticleVariable<bool> &GetMarkedForRemoval() { return marked_for_removal_; }
+
+  /// Make a new Swarm based on an existing one
+  std::shared_ptr<Swarm> AllocateCopy(const bool allocComms = false,
+                                      MeshBlock *pmb = nullptr);
+
+  /// Add variable to swarm
+  void Add(const std::string label, const Metadata &metadata);
+
+  /// Add multiple variables with common metadata to swarm
+  void Add(const std::vector<std::string> labelVector, const Metadata &metadata);
+
+  /// Remote a variable from swarm
+  void Remove(const std::string label);
+
+  /// Get real particle variable
+  ParticleVariable<Real> &GetReal(const std::string label) {
```

Comment:
make label const reference.

Reply:
Same comment here about returning ParticleVariable as a non const reference.

Reply:
Discussing with @Yurlungur keeping this as a reference maintains consistency with the grid infrastructure, and since ParticleVariable is essentially holding references to data this shouldn't have an effect on encapsulation I don't think

Reply:
So does it need to be a non const reference to be consistent with the infrastructure and not for another reason. If that's the only reason we can create an issue so it is addressed separately. If there is a reason not to return it as const beyond that then no need to do anything, @Yurlungur ?